lib/pbench/server/database/models/users.py

```diff
@@ -0,0 +1,116 @@
+import datetime
+from flask_bcrypt import generate_password_hash
+from sqlalchemy import Column, Integer, String, DateTime, LargeBinary
+from pbench.server.database.database import Database
+from sqlalchemy.orm import relationship
+
+
+class User(Database.Base):
+    """ User Model for storing user related details """
+
+    __tablename__ = "users"
+
+    id = Column(Integer, primary_key=True, autoincrement=True)
+    username = Column(String(255), unique=True, nullable=False)
+    first_name = Column(String(255), unique=False, nullable=False)
+    last_name = Column(String(255), unique=False, nullable=False)
+    password = Column(LargeBinary(128), nullable=False)
+    registered_on = Column(DateTime, nullable=False)
+    email = Column(String(255), unique=True, nullable=False)
+    auth_tokens = relationship("ActiveTokens", backref="users")
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.username = kwargs.get("username")
+        self.first_name = kwargs.get("first_name")
+        self.last_name = kwargs.get("last_name")
+        self.password = generate_password_hash(kwargs.get("password"))
+        self.email = kwargs.get("email")
+        self.registered_on = datetime.datetime.now()
+
+    def __str__(self):
+        return f"User, id: {self.id}, username: {self.username}"
+
+    @staticmethod
+    def query(**kwargs):
```

Comment:
I am not in favor of using `**kwargs` everywhere.  That prevents the interface from helping make sure the arguments to a method are as expected.

Do we all agree this is true, or are we all okay with this?

Reply:
I agree I think this is one place where I should probably not use `**kwargs`. A unique user can be queried based on `id`, `username` or `email` and we query the user object quite frequently based on the whatever is available among these 3 fields

Reply:
I would venture to say that using `**kwargs` should be rare indeed.  I think it's only required when the function is passing arguments (which it doesn't know about) through to a subroutine.  Otherwise, it really should know what its arguments are and therefore it should be using explicit parameters -- that is, it should have a well-defined interface.  Not using `**kwargs` does not mean that the function cannot receive `None` values for unspecified parameters or otherwise provide default or optional parameters/values.  And, not using `**kwargs` doesn't mean that the caller cannot provide the arguments as a (to-be-unpacked) `dict`.  So, really, AFAIK, the only real need for using `**kwargs` is when you don't know what your parameters are (or don't care), and I think that that is a rare case indeed.

Reply:
That's definitely true, but Nikhil and I were both perhaps unduly influenced by the SQLAlchemy  declarative model base class, which is `kwargs` (because the class constructor is driven by the defined schema). It was both convenient and flexible to layer on top of that. But, as Webb points out, that's not actually necessary in order to layer on top of the kwargs base class. Sigh.